core/src/test/java/io/grpc/internal/FakeClock.java

```diff
@@ -80,7 +80,9 @@ public boolean shouldAccept(Runnable command) {
       new TimeProvider() {
         @Override
         public long currentTimeNanos() {
-          return currentTimeNanos;
+          // TimeProvider uses the same clock as currentTimeMillis(), and typically doesn't have
+          // nanosecond precision.
+          return TimeUnit.MILLISECONDS.toNanos(currentTimeMillis());
```

Comment:
If you just inline `currentTimeMillis()` (on line 361) this becomes

`TimeUnit.MILLISECONDS.toNanos(TimeUnit.NANOSECONDS.toMillis(currentTimeNanos + 123456789L))`

Also there is a comment on line 361 "Normally millis and nanos are of different epochs. Add an offset to simulate that." Does any of this affect the change here?

Reply:
`System.currentTimeMillis()` is CLOCK_REALTIME, which is wall-clock/unix time in milliseconds. `System.nanoTime()` is CLOCK_MONOTONIC, which has an undefined epoch (but on Linux epoch is the boot time). `currentTimeMillis()` is for anything involving calendars and dates and specific events in time. But it can jump forward and backward in time (e.g., via `ntpdate`). `nanoTime()` doesn't jump around and so is used for timers and measuring elapsed time.

`Deadline.Ticker` is used for elapsed time, so should be `nanoTime()`. `TimeProvider` is [used for wall-time](https://github.com/grpc/grpc-java/blob/42b32c64f567a5477ad00d01c41bf3cf5b3db409/core/src/main/java/io/grpc/internal/TimeProvider.java#L22) (which you used correctly when interacting with the cert). But TimeProvider returns nanos, so it has to [scale the return value of `currentTimeMillis()`](https://github.com/grpc/grpc-java/blob/383badd21b97d99cf07741fb18761431922499a7/core/src/main/java/io/grpc/internal/TimeProvider.java#L32). That's the same behavior that I have here, which _truncates_ part of the nano time.

`TimeProvider` _used to return_ millis, [but was changed](https://github.com/grpc/grpc-java/commit/f5f560ad365735a0c9bf8b86abeed16cf7dfdde7) (and was broken because it _didn't_ jump around so [needed a fix](https://github.com/grpc/grpc-java/commit/777bbbb1218c9331ff27cd802b1c8e5d5ac5d7a5)).

Everything broke very badly with me fixing the TimeProvider here, and I don't think I'll fix it all in this PR. But the main thing in digging through the failures I realized is that the offset isn't large enough. 123_456_789 isn't even a second.

Reply:
> ...
> Everything broke very badly with me fixing the TimeProvider here, and I don't think I'll fix it all in this PR. But the main thing in digging through the failures I realized is that the offset isn't large enough. 123_456_789 isn't even a second.

Is fixing `TimeProvider` necessary for these set of tests?  I think the tests only depend on comparing successive time-stamps in calls to `checkAndReloadCertificates()` and the current system time doesn't matter. It only matters for the case of `generateErrorIfCurrentCertExpired()` which I don't think is tested in any of these tests.

Reply:
Well, it would be best if FakeClock wasn't broken. But given it is, yes, I was going to avoid trying to boil the ocean here. I've added a simple TimeProvider implementation and the changes are now relatively easy.